Move gpx test from GraphHopperWebIT to RouteResourceTest

easbar opened the pull request:
Fixes #1994 

Two bugs here:

Bug 1: Despite setting `usePostRequest` `GraphHopperWeb#export` was using GET
Bug 2: When using POST the `gpx.millis` parameter was an empty String (`timeString`) and when trying to parse it there was a `NumberFormatException` (leading to a misleading error message about an empty string)

See comments inline

easbar commented on client-hc/src/main/java/com/graphhopper/api/GraphHopperWeb.java:
```diff
@@ -64,8 +64,8 @@
     private String optimize = "false";
-    boolean postRequest = true;
-    int maxUnzippedLength = 1000;
+    private boolean postRequest = true;
```
This wasn't private before but should be now problem to do so as there is a setter anyway

easbar commented on client-hc/src/main/java/com/graphhopper/api/GraphHopperWeb.java:
```diff
@@ -64,8 +64,8 @@
-    boolean postRequest = true;
-    int maxUnzippedLength = 1000;
+    private boolean postRequest = true;
+    private int maxUnzippedLength = 1000;
```
Had to add a setter here, since we are now no longer testing this in the same package, do you think thats ok?

easbar commented on client-hc/src/main/java/com/graphhopper/api/GraphHopperWeb.java:
```diff
@@ -138,6 +138,14 @@ public GraphHopperWeb setKey(String key) {
         return this;
     }
 
+    /**
+     * Only use this if you know what you are doing
+     */
+    public GraphHopperWeb _setMaxUnzippedLength(int maxUnzippedLength) {
+        this.maxUnzippedLength = maxUnzippedLength;
+        return this;
+    }
+
```
WDYT?

easbar commented on client-hc/src/test/java/com/graphhopper/api/GraphHopperWebIT.java:
```diff
@@ -245,39 +246,6 @@ public void doNotReadFinishInstruction() {
         assertEquals("", finishInstructionName);
     }
 
-    @Test
```
I'm happy about every test we can remove here, see #1986.

easbar commented on web-bundle/src/main/java/com/graphhopper/resources/RouteResource.java:
```diff
@@ -273,7 +273,14 @@ private void removeLegacyParameters(GHRequest request) {
             throw new IllegalArgumentException("Alternatives are currently not yet supported for GPX");
         }
 
-        long time = timeString != null ? Long.parseLong(timeString) : System.currentTimeMillis();
```
This was the bug: For GET a missing `gpx.millis` would give null, but for POST it would be an empty String.

easbar commented on web-bundle/src/main/java/com/graphhopper/resources/RouteResource.java:
```diff
@@ -273,7 +273,14 @@ private void removeLegacyParameters(GHRequest request) {
+            try {
+                time = Long.parseLong(timeString);
+            } catch (NumberFormatException e) {
+                throw new IllegalArgumentException("Invalid value for 'gpx.millis': " + timeString);
```
Not handling `NumberFormatException` gave a misleading error message.

easbar commented on client-hc/src/main/java/com/graphhopper/api/GraphHopperWeb.java:
```diff
@@ -368,7 +376,7 @@ private Request createGetRequest(GHRequest ghRequest) {
         String str = "Creating request failed";
         try {
-            Request okRequest = createGetRequest(ghRequest);
+            Request okRequest = postRequest ? createPostRequest(ghRequest) : createGetRequest(ghRequest);
```
This was 'bug' Bug1

easbar commented on web/src/test/java/com/graphhopper/http/resources/RouteResourceTest.java:
```diff
@@ -502,6 +504,54 @@ public void testGPXWithError() {
+        req.putHint("type", "gpx");
+        GraphHopperWeb gh = new GraphHopperWeb(clientUrl(app, "/route"))
+                .setPostRequest(usePost)
+                ._setMaxUnzippedLength(maxUnzippedLength);
```
We could also remove maxunzipped length here since we are not testing it anyway, but we might need it for some other tests in `GraphHopperWebIT` (but then we can still add it -> removed it here :))

michaz commented:
> In `RouteResource#doGet` we have `@QueryParam("gpx.millis") String timeString)` and I was wondering first of all why is there no default value? We use `System.currMillis()` as a default later anyway. Second, why is the type String when the value is expected to be a number and we even parse it as such later? Then I tried changing to long, but in case I sent a GET request with a non-number (like `xyz`) this gives a 404 error. I assume Dropwizard interprets a query parameter that has the wrong type as an 'unknown url'. We had the same problem here: #1991, but only now I understood why we got 404. This is kind of an annoying behavior. I would prefer a 400 error stating that the parameter had the wrong type, not sure if (something in) Dropwizard has an option to change this. @michaz maybe?

When I Google 'jersey wrong query parameter type 404' I find that
a) this is by design, and
b) you're not the only one who doesn't like it

<shrug>

karussell commented:
@michaz correctly wrote that this GPX stuff is rather costly. Should we avoid including it for POST? Before releasing 1.0 would be the right time. See #2006

easbar commented:
Yeah you seem to agree that it should be removed, I do not even really know what it does :) Like I said the only thing I do not fully understand is why (if we remove it for POST) we keep it for GET, because to me GET seems to offer no advantages over POST except that it is handy to copy&paste urls of routing requests like in GH maps (for this reason I think we should keep GET, but I do not see why we need gpx for GET if it is something we can get rid of).

karussell commented:
People and customers are still using this feature and removing it from GET would mean a breaking change. At least we should give them time to migrate and do this conversion on the client.

Not introducing it in POST is inconsistent, yes, but on the other hand we do not have to add it for other endpoints that support POST?

easbar commented:
Since we removed (the anyway broken) GPX support for POST there is no longer need to fix it :) I reverted the changes here. I would still move the gpx tests to RouteResource instead of running agains graphhopper.com as I did here.

karussell commented on client-hc/src/main/java/com/graphhopper/api/GraphHopperWeb.java:
```diff
@@ -376,7 +368,8 @@ private Request createGetRequest(GHRequest ghRequest) {
         String str = "Creating request failed";
         try {
-            Request okRequest = postRequest ? createPostRequest(ghRequest) : createGetRequest(ghRequest);
+            // todonow: should there be an error if usePost=true (because it is not supported)
```
Yes, this would be probably a good idea.
